README.md
[![build status](https://secure.travis-ci.org/bda-research/node-webcrawler.png)](https://travis-ci.org/bda-research/node-webcrawler)
node-webcrawler
------------

node-webcrawler is originally a fork of [node-crawler](https://github.com/sylvinus/node-crawler) which aims to be the best crawling/scraping package for Node, sice the author won't mantain I will take care of it.

Features:
 * server-side DOM & automatic jQuery insertion with Cheerio (default) or JSDOM
 * Configurable pool size and retries
 * Priority of requests
 * forceUTF8 mode to let crawler deal for you with charset detection and conversion

Here is the [CHANGELOG](https://github.com/bda-research/node-webcrawler/blob/master/CHANGELOG.md)
 
Help & Forks welcomed!

How to install
--------------

    $ npm install node-webcrawler

Crash course
------------

```javascript
var Crawler = require("node-webcrawler");
var url = require('url');

var c = new Crawler({
    maxConnections : 10,
    // This will be called for each crawled page
    callback : function (error, result, $) {
        // $ is Cheerio by default
        //a lean implementation of core jQuery designed specifically for the server
        $('a').each(function(index, a) {
            var toQueueUrl = $(a).attr('href');
            c.queue(toQueueUrl);
        });
    }
});

// Queue just one URL, with default callback
c.queue('http://joshfire.com');

// Queue a list of URLs
c.queue(['http://jamendo.com/','http://tedxparis.com']);

// Queue URLs with custom callbacks & parameters
c.queue([{
    uri: 'http://parishackers.org/',
    jQuery: false,

    // The global callback won't be called
    callback: function (error, result) {
        console.log('Grabbed', result.body.length, 'bytes');
    }
}]);

// Queue using a function
var googleSearch = function(search) {
  return 'http://www.google.fr/search?q=' + search;
};
c.queue({
  uri: googleSearch('cheese')
});

// Queue some HTML code directly without grabbing (mostly for tests)
c.queue([{
    html: '<p>This is a <strong>test</strong></p>'
}]);
```

Options reference
-----------------

You can pass these options to the Crawler() constructor if you want them to be global or as
items in the queue() calls if you want them to be specific to that item (overwriting global options)

This options list is a strict superset of [mikeal's request options](https://github.com/mikeal/request#requestoptions-callback) and will be directly passed to
the request() method.

Basic request options:

 * `uri`: String, the URL you want to crawl
 * `timeout` : Number, in milliseconds        (Default 15000)
 * [All mikeal's requests options are accepted](https://github.com/mikeal/request#requestoptions-callback)

Callbacks:

 * `callback(error, result, $)`: A request was completed
 * `onDrain(pool)`: There is no more queued requests, call `pool.destroyAllNow()` if you wanna release resources in pool to, or if you have follow-up tasks to queue you can ignore.

Pool options:

 * `maxConnections`: Number, Size of the worker pool (Default 10),
 * `priorityRange`: Number, Range of acceptable priorities starting from 0 (Default 10),
 * `priority`: Number, Priority of this request (Default 5),

Retry options:

 * `retries`: Number of retries if the request fails (Default 3),
 * `retryTimeout`: Number of milliseconds to wait before retrying (Default 10000),

Server-side DOM options:

 * `jQuery`: true, false or ConfObject (Default true)

Charset encoding:

 * `forceUTF8`: Boolean, if true will get charset from HTTP headers or meta tag in html and convert it to UTF8 if necessary. Never worry about encoding anymore! (Default false),
 * `incomingEncoding`: String, with forceUTF8: true to set encoding manually (Default null)
     `incomingEncoding : 'windows-1255'` for example

Cache:

 * `cache`: Boolean, if true stores requests' result in memory (Default false), not recommend if you are doing with huge amount of pages as the process will exhaust momery
 * `skipDuplicates`: Boolean, if true skips URIs that were already crawled, without even calling callback() (Default false)

Other:
 * `rotateUA`: Boolean, if true, `userAgent` should be an array, and rotate it (Default false) 
 * `userAgent`: String or Array, if `rotateUA` is false, but `userAgent` is array, will use first one. defaults to "node-crawler/[version]"
 * `referer`: String, if truthy sets the HTTP referer header
 * `rateLimits`: Number of milliseconds to delay between each requests (Default 0) 

 
Class:Crawler
-------------

Instance of Crawler

__crawler.queue(uri|options)__
 * `uri` String, `options` is [Options](#options-reference)

__crawler.queueSize__

Size of queue, read-only

 
Working with Cheerio or JSDOM
-----------------------------

Crawler by default use [Cheerio](https://github.com/cheeriojs/cheerio) instead of [Jsdom](https://github.com/tmpvar/jsdom). Jsdom is more robust but can be hard to install (espacially on windows) because of [contextify](https://github.com/tmpvar/jsdom#contextify).
Which is why, if you want to use jsdom you will have to build it, and `require('jsdom')` in your own script before passing it to crawler. This is to avoid cheerio crawler user to build jsdom when installing crawler.

###Working with Cheerio
```javascript
jQuery: true //(default)
//OR
jQuery: 'cheerio'
//OR
jQuery: {
    name: 'cheerio',
    options: {
        normalizeWhitespace: true,
        xmlMode: true
    }
}
```
These parsing options are taken directly from [htmlparser2](https://github.com/fb55/htmlparser2/wiki/Parser-options), therefore any options that can be used in `htmlparser2` are valid in cheerio as well. The default options are:

```js
{
    normalizeWhitespace: false,
    xmlMode: false,
    decodeEntities: true
}
```

For a full list of options and their effects, see [this](https://github.com/fb55/DomHandler) and
[htmlparser2's options](https://github.com/fb55/htmlparser2/wiki/Parser-options).
[source](https://github.com/cheeriojs/cheerio#loading)

###Working with JSDOM

In order to work with JSDOM you will have to install it in your project folder `npm install jsdom`, deal with [compiling C++](https://github.com/tmpvar/jsdom#contextify) and pass it to crawler.
```javascript
var jsdom = require('jsdom');
var Crawler = require('node-webcrawler');

var c = new Crawler({
    jQuery: jsdom
});
```

How to test
-----------

### Install and run Httpbin

node-webcrawler use a local httpbin for testing purpose. You can install httpbin as a library from PyPI and run it as a WSGI app. For example, using Gunicorn:

    $ pip install httpbin
    // launch httpbin as a daemon with 6 worker on localhost
    $ gunicorn httpbin:app -b 127.0.0.1:8000 -w 6 --daemon

    // Finally
    $ npm install && npm test

### Alternative: Docker

After [installing Docker](http://docs.docker.com/), you can run:

    // Builds the local test environment
    $ docker build -t node-webcrawler .

    // Runs tests
    $ docker run node-webcrawler sh -c "gunicorn httpbin:app -b 127.0.0.1:8000 -w 6 --daemon && npm install && npm test"

    // You can also ssh into the container for easier debugging
    $ docker run -i -t node-webcrawler bash

    
[![build status](https://secure.travis-ci.org/bda-research/node-webcrawler.png)](https://travis-ci.org/bda-research/node-webcrawler)

Rough todolist
--------------

 * Using bottleneck to deal with rate limits
 * Introducing zombie to deal with page with complex ajax
 * Refactoring the code to be more maintenable, it's spaghetti code in there !
 * Proxy feature
 * This issue: https://github.com/sylvinus/node-crawler/issues/118
 * Make Sizzle tests pass (jsdom bug? https://github.com/tmpvar/jsdom/issues#issue/81)
 * More crawling tests
 * Document the API more (+ the result object)
 * Option to wait for callback to finish before freeing the pool resource (via another callback like next())


ChangeLog
---------

See https://github.com/bda-research/node-webcrawler/releases


CHANGELOG.md
node-webcrawler ChangeLog
-------------------------

0.5.0 version changelog:
 * parse charset from `content-type` in http headers or meta tag in html, then convert
 * big5 charset is avaliable as the `iconv-lite` has already supported it 
 * default enable gzip in request header
 * remove unzip code in crawler since `request` will do this
 * body will return as a Buffer if encoding is null which is an option in `request`
 * remove cache and skip duplicate `request` for `GET`, `POST`(only for type `urlencode`), `HEAD`
 * add log feature, you can use `winston` to set `logger:winston`, or crawler will output to console
 * rotate user-agent in case some sites ban your requests
 
0.5.1 version changelog:
 * remove cache feature, it's useless
 * add `localAddress`, `time`, `tunnel`, `proxyHeaderWhiteList`, `proxyHeaderExclusiveList` properties to pass to `request`

0.5.2 version changelog:
 * you can manually terminate all the resources in your pool, when `onDrain` called, before their timeouts have been reached
 * add a read-only property `queueSize` to crawler
 
0.6.0 version changelog:
 * add `bottleneck` to implement rate limit, one can set limit for each connection at same time.
 
0.6.3 version changelog:
 * you could also get `result.options` from callback even when some errors ouccurred
 * add test for `bottleneck`

0.6.5
 * fix a deep and big bug when initializing Pool, that may lead to sequence execution. [issue](https://github.com/bda-research/node-webcrawler/issues/2)
 * print log of Pool status


lib/crawler.js
'use strict';

var path = require('path');
var util = require('util');
var EventEmitter = require('events').EventEmitter;

var request = require('request');
var _ = require('lodash');
var cheerio = require('cheerio');
var fs = require('fs');
var Pool = require('generic-pool').Pool;
var charsetParser = require('charset-parser');
var Bottleneck = require('bottleneck');
var seenreq = require('seenreq');
var logger=null;
// Fallback on iconv-lite if we didn't succeed compiling iconv
// https://github.com/sylvinus/node-crawler/pull/29
var iconv, iconvLite;
try {
    iconv = require('iconv').Iconv;
} catch (e) {}

if (!iconv) {
    iconvLite = require('iconv-lite');
}

function useCache (options) {
    return (options.uri || options.url) && options.cache;
}

function checkJQueryNaming (options) {
    if ('jquery' in options) {
        options.jQuery = options.jquery;
        delete options.jquery;
    }
    return options;
}

function readJqueryUrl (url, callback) {
    if (url.match(/^(file\:\/\/|\w+\:|\/)/)) {
        fs.readFile(url.replace(/^file\:\/\//,''),'utf-8', function(err,jq) {
            callback(err, jq);
        });
    } else {
        callback(null, url);
    }
}

function Crawler (options) {
    var self = this;
    self.init(options);
}
// augment the prototype for node events using util.inherits
util.inherits(Crawler, EventEmitter);

Crawler.prototype.init = function init (options) {
    var self = this;

    var defaultOptions = {
        autoWindowClose:    true,
        cache:              false,
        forceUTF8:          false,
	gzip:               true,
        incomingEncoding:   null, //TODO remove or optimize
        jQuery:             true,
        maxConnections:     10,
        method:             'GET',
        onDrain:            false,
        priority:           5,
        priorityRange:      10,
        rateLimits:         0,
        referer:            false,
        retries:            3,
        retryTimeout:       10000,
	timeout:            15000,
        skipDuplicates:     false,
	rotateUA:           false
    };

    //return defaultOptions with overriden properties from options.
    self.options = _.extend(defaultOptions, options);

    // you can use jquery or jQuery
    self.options = checkJQueryNaming(self.options);

    if (self.options.rateLimits !== 0 && self.options.maxConnections == 1) {
	//self.options.limiter = "default";
    }else{
	//self.limiters = null;//self.options.maxConnections = 1;
    }
    
    // Don't make these options persist to individual queries
    self.globalOnlyOptions = ['maxConnections', 'priorityRange', 'onDrain'];

    //Setup a worker pool w/ https://github.com/coopernurse/node-pool
    self.pool = Pool({
        name         : 'crawler',
        max          : self.options.maxConnections,
	min          : self.options.minConnections,
	log          : self.options.logger && function(){self.options.logger.log(arguments[1],arguments[0]);},
        priorityRange: self.options.priorityRange,
        create       : function(callback) {
            callback(new Object());
        },
        destroy      : function() {}
    });
    
    self.limiters = new Bottleneck.Cluster(self.options.maxConnections,self.options.rateLimits);
    self.plannedQueueCallsCount = 0;
    self.queueItemSize = 0;

    self.cache = {};
    self.seen = new seenreq();
    logger = self.options.logger || console;

    self.on('pool:release', function(options) {
        self._release(options);
    });
    
    self.on("request",function(options){
	
	if(_.isFunction(self.options.preRequest)){
	    self.options.preRequest(options);
	}
    });
    
    self.on('pool:drain', function() {
        if (self.options.onDrain) {
            self.options.onDrain(self.pool);
        }
    });
};

Crawler.prototype._release = function _release (options) {
    var self = this;

    self.queueItemSize--;
    if (options._poolReference) {
	if(options.debug){
	    logger.info("Releasing resource, %s",options.limiter);
	}
	
        self.pool.release(options._poolReference);
    }

    // Pool stats are behaving weird - have to implement our own counter
    if (self.queueItemSize + self.plannedQueueCallsCount === 0) {
        self.emit('pool:drain');
    }
};

Crawler.prototype._inject = function _inject (response, options, callback) {
    var $;
    var self = this;

    if (options.jQuery === 'cheerio' || options.jQuery.name === 'cheerio' || options.jQuery === true) {
        var defaultCheerioOptions = {
            normalizeWhitespace: false,
            xmlMode: false,
            decodeEntities: true
        };
        var cheerioOptions = options.jQuery.options || defaultCheerioOptions;
        $ = cheerio.load(response.body, cheerioOptions);

        callback(null, $);
    }

    else if (options.jQuery.jsdom) {
        var jsdom = options.jQuery.jsdom;
        var scriptLocation = path.resolve(__dirname, '../vendor/jquery-2.1.1.min.js');

        //Use promises
        readJqueryUrl(scriptLocation, function(err, jquery) {
            try {
                jsdom.env({
                    url: options.uri,
                    html: response.body,
                    src: [jquery],
                    done: function (errors, window) {
                        $ = window.jQuery;
                        callback(errors, $);

                        try {
                            window.close();
                            window = null;
                        } catch (err) {
                            logger.error(err);
                        }

                    }
                });
            } catch (e) {
                options.callback(e);
                self.emit('pool:release', options);
            }
        });
    }
    // Jquery is set to false are not set
    else {
        callback(null);
    }
};

Crawler.prototype.queue = function queue (options) {
    var self = this;
    
    // Did you get a single object or string? Make it compatible.
    options = _.isString(options) || _.isPlainObject(options) ? [ options ] : options;
    if (_.isArray(options) && options.length == 1) {
        self._pushToQueue(
            _.isString(options[0]) ? { uri: options[0] } : options[0]
        );
    // Did you get multiple requests? Queue the URLs.
    } else if (!_.isNull(options) && !_.isUndefined(options) && options.length>0) {
        self.queue(
            _.isString(options[0]) ? { uri: options[0] } : options[0]
        );
        self.queue(options.slice(1))
    }
};

Crawler.prototype._pushToQueue = function _pushToQueue (options) {
    var self = this;
    self.queueItemSize++;

    // you can use jquery or jQuery
    options = checkJQueryNaming(options);

    _.defaults(options, self.options);

    // Remove all the global options from our options
    // TODO we are doing this for every _pushToQueue, find a way to avoid this
    _.each(self.globalOnlyOptions, function(globalOnlyOption) {
        delete options[globalOnlyOption];
    });

    // If duplicate skipping is enabled, avoid queueing entirely for URLs we already crawled
    if (options.skipDuplicates && self.seen.exists(options)) {
        return self.emit('pool:release', options);
    }

    // acquire connection - callback function is called
    // once a resource becomes available
    // self.pool.acquire(
    var acquired = function(error, poolReference) {
        options._poolReference = poolReference;

        // this is and operation error
        if (error) {
            logger.error(error);
            options.callback(error);// need release
	    return self.emit('pool:release',options);
        }

	if(options.debug){
	    logger.info("Acquired resource, %s",options.limiter);
	}
	
        //Static HTML was given, skip request
        if (options.html) {
            self._onContent(null, options, {body:options.html});
        } else if (typeof options.uri === 'function') {
            options.uri( function(uri) {
                options.uri = uri;
                self._makeCrawlerRequest(options);
            });
        } else {
            self._makeCrawlerRequest(options);
        }
    }//, options.priority);
    
    var acquireWrapped = function(priority,cb){
	if(options.debug){
	    logger.info("Called by bottleneck, %s",options.limiter);
	}

    	return self.pool.acquire(cb,priority);
    }
    
    self.limiters.key(options.limiter||"default").submit(acquireWrapped,options.priority,acquired);
};

Crawler.prototype._makeCrawlerRequest = function _makeCrawlerRequest (options) {
    var self = this;
    //var cacheData = self.cache[self.seen.normalize(options)];
    
    // if(useCache(options) && cacheData){
    // 	if(options.debug){
    // 	    logger.info("using cache.");
    // 	}
	
    // 	self._onContent(null, options, cacheData, true);
    // 	return;
    // }
    
    // if (typeof options.rateLimits === 'number' && options.rateLimits !== 0) {
    //     setTimeout(function() {
    //         self._buildHttpRequest(options);
    //     }, options.rateLimits);
    // } else {
        self._buildHttpRequest(options);
    // }
};

Crawler.prototype._buildHttpRequest = function _buildHTTPRequest (options) {
    var self = this;

    if (options.debug) {
        logger.info(options.method+' '+options.uri);
    }

    // Cloning keeps the opts parameter clean:
    // - some versions of "request" apply the second parameter as a
    // property called "callback" to the first parameter
    // - keeps the query object fresh in case of a retry
    // Doing parse/stringify instead of _.clone will do a deep clone and remove functions

    var ropts = JSON.parse(JSON.stringify(options));

    if (!ropts.headers) { ropts.headers={}; }
    if (ropts.forceUTF8) {
        // if (!ropts.headers['Accept-Charset'] && !ropts.headers['accept-charset']) {
        //     ropts.headers['Accept-Charset'] = 'utf-8;q=0.7,*;q=0.3';
        // }
	
        ropts.encoding=null;
    }
    
    if (ropts.userAgent) {
	if(ropts.rotateUA && _.isArray(ropts.userAgent)){
	    ropts.headers['User-Agent'] = ropts.userAgent[0];
	    // If "rotateUA" is true, rotate User-Agent
	    options.userAgent.push(options.userAgent.shift());
	}else{
            ropts.headers['User-Agent'] = ropts.userAgent;
	}
	if(options.debug){
	    logger.info(ropts.headers['User-Agent']);
	}
    }
    if (ropts.referer) {
        ropts.headers.Referer = ropts.referer;
    }
    if (ropts.proxies && ropts.proxies.length) {
        ropts.proxy = ropts.proxies[0];
    }
    
    this.emit("request",ropts);

    var requestArgs = ['uri','url','qs','method','headers','body','form','json','multipart','followRedirect',
        'followAllRedirects', 'maxRedirects','encoding','pool','timeout','proxy','auth','oauth','strictSSL',
        'jar','aws','gzip','time','tunnel','proxyHeaderWhiteList','proxyHeaderExclusiveList','localAddress'];

    var req = request(_.pick.apply(this,[ropts].concat(requestArgs)), function(error,response) {
        if (error) {
            return self._onContent(error, options);
        }
	
        response.uri = response.request.href;
        self._onContent(error,options,response);
    });
};

Crawler.prototype._onContent = function _onContent (error, options, response, fromCache) {
    var self = this;

    if (error) {
        if (options.debug) {
            logger.error('Error '+error+' when fetching '+
            options.uri+(options.retries?' ('+options.retries+' retries left)':''));
        }
        if (options.retries) {
            self.plannedQueueCallsCount++;
            setTimeout(function() {
                options.retries--;
                self.plannedQueueCallsCount--;

                // If there is a "proxies" option, rotate it so that we don't keep hitting the same one
                // if (options.proxies) {
                //     options.proxies.push(options.proxies.shift());
                // }
		self.queue(options);
            },options.retryTimeout);

        } else if (options.callback) {
            options.callback(error,{options:options});
        }

        return self.emit('pool:release', options);
    }
    
    if (!response.body) { response.body=''; }

    if (options.debug) {
        logger.info('Got '+(options.uri||'html')+' ('+response.body.length+' bytes)...');
    }

    if(!fromCache){
	try{
            self._doEncoding(options,response);
	}catch(e){
	    logger.error(e);
	    if(options.callback){
		options.callback(e);
	    }
	    return self.emit('pool:release',options);
	}
    }
    
    // if(useCache(options)){
    // 	self.cache[self.seen.normalize(options)] = response;
    // }
    
    if (!options.callback) {
        return self.emit('pool:release', options);
    }

    response.options = options;

    // This could definitely be improved by *also* matching content-type headers
    var isHTML = _.isString(response.body) && response.body.match(/^\s*</);

    if (isHTML && options.jQuery && options.method !== 'HEAD') {
        self._inject(response, options, function(errors, $) {
            self._onInject(errors, options, response, $);
        });
    } else {
        options.callback(null,response);
        self.emit('pool:release', options);
    }
};

Crawler.prototype._doEncoding = function(options,response){
    var self = this;
    
    if(options.encoding === null){
	return;
    }
    
    if (options.forceUTF8) {
        var iconvObj;
	var charset = options.incomingEncoding || self._parseCharset(response);
	
	if (options.debug) {
            logger.info('Charset ' + charset);
        }
	
        if (charset !== 'utf-8' && charset !== 'ascii') {
            if (iconv) {
                iconvObj = new iconv(charset, 'UTF-8//TRANSLIT//IGNORE');
		response.body = iconvObj.convert(response.body).toString();
            } else{
                response.body = iconvLite.decode(response.body, charset);
            }
        }
    }
    
    //if charset = 'utf-8', call toString() ;
    response.body = response.body.toString();
}

Crawler.prototype._onInject = function _onInject (errors, options, response, $) {
    var self = this;
    
    options.callback(errors, response, $);
    self.emit('pool:release', options);
};

Crawler.prototype._parseCharset = function(res){
    var ctt = res.headers['content-type']||res.headers['Content-Type']||'';
    var body = res.body instanceof Buffer?res.body.toString():res.body;
    var charset = charsetParser(ctt,body,'utf-8');
    
    return charset;
}

Object.defineProperty(Crawler.prototype,'queueSize',{
    get:function(){
	return this.pool.waitingClientsCount();
    }
})

Object.defineProperty(Crawler.prototype,'waitingCount',{
    get:function(){
	return this.pool.waitingClientsCount();
    }
})

Object.defineProperty(Crawler.prototype,'availableCount',{
    get:function(){
	return this.pool.availableObjectsCount();
    }
})

Object.defineProperty(Crawler.prototype,'size',{
    get:function(){
	return this.pool.getPoolSize();
    }
})

module.exports = Crawler;
module.exports.VERSION = '0.6.4';
